compute/resource_pipeline.go

```diff
@@ -0,0 +1,228 @@
+	AllowDuplicateNames bool              `json:"allow_duplicate_names,omitempty"`
+}
+
+type createPipelineResponse struct {
+	PipelineID string `json:"pipeline_id"`
+}
+
+// PipelineState ...
+type PipelineState string
+
+// Constants for PipelineStates
+const (
+	StateDeploying  PipelineState = "DEPLOYING"
+	StateStarting   PipelineState = "STARTING"
+	StateRunning    PipelineState = "RUNNING"
+	StateStopping   PipelineState = "STOPPPING"
+	StateDeleted    PipelineState = "DELETED"
+	StateRecovering PipelineState = "RECOVERING"
+	StateFailed     PipelineState = "FAILED"
+	StateResetting  PipelineState = "RESETTING"
+	StateIdle       PipelineState = "IDLE"
+)
+
+// PipelineHealthStatus ...
+type PipelineHealthStatus string
+
+// Constants for PipelineHealthStatus
+const (
+	HealthStatusHealthy   PipelineHealthStatus = "HEALTHY"
+	HealthStatusUnhealthy PipelineHealthStatus = "UNHEALTHY"
+)
+
+type pipelineInfo struct {
+	PipelineID string                `json:"pipeline_id"`
+	Spec       *pipelineSpec         `json:"spec"`
+	State      *PipelineState        `json:"state"`
+	Cause      string                `json:"cause"`
+	ClusterID  string                `json:"cluster_id"`
+	Name       string                `json:"name"`
+	Health     *PipelineHealthStatus `json:"health"`
+}
+
+type pipelinesAPI struct {
+	client *common.DatabricksClient
+	ctx    context.Context
+}
+
+func newPipelinesAPI(ctx context.Context, m interface{}) pipelinesAPI {
+	return pipelinesAPI{m.(*common.DatabricksClient), ctx}
+}
+
+func (a pipelinesAPI) create(s pipelineSpec, allowDuplicateNames bool) (string, error) {
+	var id createPipelineResponse
+	err := a.client.Post(a.ctx, "/pipelines", s, &id)
+	return id.PipelineID, err
+}
+
+func (a pipelinesAPI) read(id string) (p pipelineInfo, err error) {
+	err = a.client.Get(a.ctx, "/pipelines/"+id, nil, &p)
+	return
+}
+
+func (a pipelinesAPI) update(id string, s pipelineSpec, allowDuplicateNames bool) error {
+	return a.client.Put(a.ctx, "/pipelines/"+id, s)
+}
+
+func (a pipelinesAPI) delete(id string) error {
+	return a.client.Delete(a.ctx, "/pipelines/"+id, map[string]string{})
+}
+
+func (a pipelinesAPI) defaultTimeout() time.Duration {
+	return 30 * time.Minute
+}
+
+func (a pipelinesAPI) retryFunc(id string, desiredState PipelineState, lastState *PipelineState) *resource.RetryError {
+	s, err := a.read(id)
+	if err != nil {
+		if e, ok := err.(common.APIError); ok && e.ErrorCode == "RESOURCE_DOES_NOT_EXIST" {
+			*lastState = StateDeleted
+		} else {
+			return resource.NonRetryableError(err)
+		}
+	} else {
+		lastState = s.State
+	}
+	log.Printf("[DEBUG] Pipeline %s is in state %s", id, *lastState)
+	if *lastState == desiredState {
+		return nil
+	}
+	return resource.RetryableError(fmt.Errorf("Pipeline %s is in state %s, not yet in state %s", id, *lastState, desiredState))
+}
+
+func (a pipelinesAPI) waitForState(id string, desiredState PipelineState) (PipelineState, error) {
+	var lastState PipelineState
```

Comment:
please remove this lastState, as you're ignoring the result anyway :) and please merge retryFunc and waitForState - it would be more readable that way. 

please also add `timeout time.Duration` argument, as you'll have to propagate down a user-specified timeout as `d.Timeout(schema.TimeoutCreate)`. System default is 20 minutes. You can look at `ResourceWorkspace` for an example. That one is probably the most readable, along with sql endpoint.